### ISSUE TITLE
Train Model 4: Reranker (5M params)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#668]
## Goal

Train a learned model to rerank RAG retrieval results by relevance to query, improving retrieval quality beyond BM25.

## Architecture

```python
class SentenceReranker(nn.Module):
    # 5M parameters
    # Bidirectional LSTM sentence encoder (256 hidden, 2 layers)
    # Query encoder
    # Relevance scorer
```

## What It Does

- Takes top-K (e.g., 100) from initial retrieval
- Reranks to top-N (e.g., 20) by semantic relevance
- Uses AST structure + compositional embeddings

## When It Helps

```
Query: "Kiu fondis Esperanton?"

Sentence 1: "Zamenhof fondis Esperanton en 1887."
BM25: 0.82 → Reranker: 0.95 ✅ (highly relevant)

Sentence 2: "Esperanto havas regulan gramatikon."
BM25: 0.45 → Reranker: 0.15 ❌ (not relevant to WHO)

Sentence 3: "Zamenhof estis pola kuracisto."
BM25: 0.38 → Reranker: 0.75 ✅ (answers WHO - person identification)
```

## Training Data Needed

10,000-20,000 examples: (query, sentence, relevance_label)

**Relevance labels**: 0 (not relevant), 0.5 (partially), 1.0 (highly)

**Collection strategies**:
1. Manual annotation (expensive but accurate)
2. Weak supervision (sentences used in human summaries = relevant)
3. Contrastive learning (same-document sentences = positive pairs)

## Training Approach

```python
# Option 1: Pairwise ranking loss
loss = max(0, margin - (pos_score - neg_score))

# Option 2: Binary cross-entropy
loss = nn.BCELoss()(predicted_relevance, gold_relevance)
```

## Quality Metrics

- Precision@10: >0.80 (80% of top-10 are relevant)
- Recall@20: >0.90 (90% of relevant sentences in top-20)
- Improved summary quality: +10-15% over BM25-only

## Dependencies

- Root Embeddings (#479, should be retrained first)
- M1 Selectional (#475) - used before reranker in pipeline
- RAG retrieval infrastructure

## Timeline

2 weeks (Phase 2, Week 1-2)

## Related

- See: docs/COMPLETE_SYSTEM_DESIGN_WITH_MODELS.md (Model 4)
- Part of Epic #654

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
